### PR TITLE
Dev to main

### PR DESCRIPTION
**Pull request by kw4ntiq**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Norm of the code has been respected

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of change does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **What is the current behavior?** (link an issue based on the kind of change this pr introduce)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

**Review by KitetsuK — APPROVED** [review]

**Review by Saverio976 — APPROVED** [review]

**Review by romainpanno — APPROVED** [review]

**Review by guillaumeAbel — APPROVED** [review]

File: src/Nitwork/ANitwork.cpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            Logger::fatal("NITWORK: catch clock thread: not joinable");
+            return false;
+        }
+        return true;
+    }
+
+    bool ANitwork::startNitworkThreads(int threadNb, int tick)
+    {
+        if (!startContextThreads(threadNb)) {
+            std::cerr << "Error: context threads failed" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/ANitwork.cpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            return false;
+        }
+        if (!startClockThread(tick)) {
+            std::cerr << "Error: clock thread failed" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/ANitwork.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    void ANitwork::callReceiveHandler(const std::string &message)
+    {
+        Logger::error(message);
+        startReceiveHandler();
+    }
+
+    void ANitwork::headerHandler(std::size_t bytes_received, const boost::system::error_code &error)
+    {
+        if (error) {
+            callReceiveHandler("Error: " + error.message());
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]:
```suggestion
            callReceiveHandler("NITWORK: " + error.message());
```

File: src/Nitwork/ANitwork.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            return;
+        }
+        if (bytes_received < sizeof(struct header_s)) {
+            callReceiveHandler("Error: header not received");
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]:
```suggestion
            callReceiveHandler("NITWORK: header not received");
```

File: src/Nitwork/ANitwork.cpp [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            return;
+        }
+        // NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto *header = reinterpret_cast<struct header_s *>(_receiveBuffer.data());
+        // NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast)
+
+        if (header->magick1 != HEADER_CODE1 || header->magick2 != HEADER_CODE2) {
+            callReceiveHandler("Error: header magick not valid");
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]:
```suggestion
            callReceiveHandler("NITWORK: header magick not valid");
```

File: src/Nitwork/ANitwork.cpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            return;
+        }
+        if (header->nb_action > MAX_NB_ACTION || header->nb_action < 0 || isAlreadyReceived(header->id)) {
+            callReceiveHandler("Error: header nb action not valid or already received");
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]:
```suggestion
            callReceiveHandler("NITWORK: header nb action not valid or already received");
```

File: src/Nitwork/ANitwork.cpp [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        }
+        std::reverse(ids.begin(), ids.end());
+        for (std::size_t index = 0; index < _packetsSent.size(); index++) {
+            if (ids[index] == 0) {
+                auto packet =
+                    std::find_if(_packetsSent.begin(), _packetsSent.end(), [header, index](auto &packet) {
+                        return std::size_t(packet.second.id) == header.last_id_received - index;
+                    });
+                if (packet == _packetsSent.end()) {
+                    std::cerr << "Error: packet not found" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/ANitwork.cpp [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                        return a.first.id < b.first.id;
+                    });
+                    for (auto &action : _actions) {
+                        action.second(action.first.data, action.first.endpoint);
+                    }
+                    _actions.clear();
+                    _inputQueueMutex.unlock();
+                }
+            } catch (std::exception &e) {
+                std::cerr << e.what() << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/ANitwork.cpp [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            }
+        });
+    }
+
+    void ANitwork::sendPackages(const std::map<enum n_actionType_t, actionHandler> &actionToSendHandlers)
+    {
+        for (auto &data : _outputQueue) {
+            auto it = actionToSendHandlers.find(data.second.action);
+            if (it == actionToSendHandlers.end()) {
+                std::cerr << "Error: action not found" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/ANitwork.cpp [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+                    _outputQueueMutex.lock();
+                    _outputQueue.sort([](auto &a, auto &b) {
+                        return a.second.id < b.second.id;
+                    });
+                    sendPackages(actionToSendHandlers);
+                    _outputQueue.clear();
+                    _outputQueueMutex.unlock();
+                }
+            } catch (std::exception &e) {
+                std::cerr << e.what() << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/ANitwork.cpp]: logger

File: src/Nitwork/NitworkClient.cpp [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        const struct header_s &header,
+        const boost::asio::ip::udp::endpoint &endpoint)
+    {
+        // NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast,
+        // cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        auto *action = reinterpret_cast<struct action_s *>(_receiveBuffer.data() + sizeof(struct header_s));
+        // NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast,
+        // cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        if (endpoint.address().to_string() != _endpoint.address().to_string()) {
+            std::cerr << "Error: endpoint is not the server" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/NitworkClient.cpp]: logger

File: src/Nitwork/NitworkClient.cpp [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+            return;
+        }
+        auto it = _actionsHandlers.find(action->magick);
+        if (it == _actionsHandlers.end()) {
+            std::cerr << "Error: action not found" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/NitworkClient.cpp]: logger

File: src/Nitwork/NitworkClient.cpp [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+            return;
+        }
+        Logger::info(
```
Comment (Saverio976) [inline src/Nitwork/NitworkClient.cpp]:
```suggestion
        Logger::debug(
```

File: src/Nitwork/NitworkClient.cpp [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+    /* Handlers Section */
+    // NOLINTBEGIN(readability-convert-member-functions-to-static)
+    void NitworkClient::handleStartGame(const std::any &msg, boost::asio::ip::udp::endpoint & /* unused */)
+    {
+        // NOLINTEND(readability-convert-member-functions-to-static)
+        const struct msgStartGame_s &msgStartGame = std::any_cast<struct msgStartGame_s>(msg);
+
+        if (msgStartGame.magick != MAGICK_START_GAME) {
+            std::cerr << "Error: magick is not START_GAME" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/NitworkClient.cpp]: logger

File: src/Nitwork/NitworkClient.cpp [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+            return;
+        }
+        std::cout << "Game started" << std::endl;
```
Comment (Saverio976) [inline src/Nitwork/NitworkClient.cpp]: logger

**Review by Saverio976 — DISMISSED** [review]

File: src/Nitwork/NitworkServer.cpp [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                         .nb_action        = 1,
+                         .magick2          = HEADER_CODE2},
+            .action = {.magick = INIT},
+            .msg    = {.magick = MAGICK_INIT, .playerId = playerId}
+        };
+        Packet packet(
+            packetMsgPlayerInit.header.id,
+            packetMsgPlayerInit.action.magick,
+            std::make_any<struct packetMsgPlayerInit_s>(packetMsgPlayerInit));
+        std::cout << "Send PLAYER_INIT to " << endpoint.address().to_string() << ":" << endpoint.port()
```
Comment (guillaumeAbel) [inline src/Nitwork/NitworkServer.cpp]: cout

File: src/Nitwork/ANitwork.hpp [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+            }
+
+            template <typename T>
+            Packet updateHeaderPacket(const Packet &packet)
+            {
+                T data = std::any_cast<T>(packet.body);
+
+                data.header.ids_received = getIdsReceived();
+                auto updatedPacket       = Packet(packet.id, packet.action, std::make_any<T>(data));
+                std::cout << "updatedPacket.header.ids_received: " << data.header.ids_received << std::endl;
```
Comment (guillaumeAbel) [inline src/Nitwork/ANitwork.hpp]: cout

File: src/Nitwork/NitworkServer.cpp [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+             getPacketID(),
+             1, HEADER_CODE2},
+            {START_GAME},
+            {MAGICK_START_GAME, playerId}
+        };
+        Packet packet(
+            packetMsgStartGame.header.id,
+            packetMsgStartGame.action.magick,
+            std::make_any<struct packetMsgStartGame_s>(packetMsgStartGame));
+        std::cout << "Send START_GAME to " << endpoint.address().to_string() << ":" << endpoint.port()
```
Comment (guillaumeAbel) [inline src/Nitwork/NitworkServer.cpp]: cout

**Review by guillaumeAbel — CHANGES_REQUESTED** [review]

**Review by guillaumeAbel — APPROVED** [review]

**Review by romainpanno — APPROVED** [review]

**Pull request merged**